keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,537 @@
+
+## Motivation
+
+The volume snapshot feature allows creating/deleting volume snapshots, and the ability to create new volumes from a snapshot natively using the Kubernetes API. However, application consistency is not guaranteed. An user has to figure out how
+to quiece an application before taking a snapshot and unquiece it after taking the snapshot.
+
+So we want to introduce an `ExecutionHook` to facilitate the quiesce and unquiesce actions when taking a snapshot. There is an existing lifecycle hook in the `Container` struct. The lifecycle hook is called immediately after a container is created or immediately before a container is terminated. The proposed execution hook is not tied to the start or termination time of the container. It can be triggered on demand by callers (users or controllers) and the status will be updated dynamically.
+
+### Goals
+
+`ExecutionHook` is introduced to define actions that can be taken on a container or a group of containers. The ExecutionHook controller is responsible for triggering the commands in containers and updating the status on whether the execution is succeeded or not. The controller will also garbage collect the hook objects based on some predefined policy.
+
+### Non-Goals
+
+This proposal does not provide exact command included in the `ExecutionHook`
+because every application has a different requirement.
+
+ApplicationSnapshot and GroupSnapshot will be mentioned in this proposal whenever relevant, but detailed design will not be included in this spec.
+
+## Proposal
+
+The general ExecutionHook API has two parts, spec and status. The hook spec has two piece of information, what are the commands to execute and where to execute them. In many use cases, different hooks share the same commands and user or controller will create many hooks for execution repeatedly. To reduce the work of copying the same execution commands in different hooks, we also propose a second API, HookAction to record the execution commands which can be referenced by ExecutionHook API. Both APIs are namespaced and they should be in the same namespace.
+
+The Create event of ExecutionHook will trigger the hook controller to run HookAction command. The status will be updated once it is available by controller. It is the caller's (user or controller) responsibility to delete the hook once the execution finishes (either succeeded or failed). Otherwise, ExecutionHook controller will garbage collect them after a certain amount of time.
+
+Here is the definition of an ExecutionHook API object:
+
+```
+// ExecutionHook is in the tenant namespace
+type ExecutionHook struct {
```

Comment:
From a non-code comment:

Let's talk about names. Looking forward to it being in Pod, I would propose "actions" or "notifications". You seem to like action, I am ok with that for now.

Since we want to keep the declarative request for action, I propose to rename ExecutionHook -> PodAction

Since we want to move the definition of the action itself into Pod, the name matters less. Rename HookAction -> PodActionTemplate or PodActionDefinition or something like that?